src/dmd/cparse.d

```diff
@@ -2266,6 +2267,38 @@ final class CParser(AST) : Parser!AST
                     break;
                 }
 
+                case TOK.typeof_:
+                {
+                    nextToken();
+                    check(TOK.leftParenthesis);
+
+                    void typeofExpr()
+                    {
+                        auto e = cparseExpression();
+                        auto ie = e.isIdentifierExp();
+                        auto td = !ie ? null : lookupTypedef(ie.ident);
+                        if(td && *td)
+                            t = *td;
+                        else
+                            t = new AST.TypeTypeof(token.loc, e);
+                    }
+
+                    auto tk = &token;
+                    if (isPrimaryExpression(tk))
+                        typeofExpr();
+                    // must be checked between primary/assignment lest type names instead parse as symbols
+                    else if (isTypeName(tk))
+                        t = cparseTypeName();
+                    else if (isAssignmentExpression(tk))
+                        typeofExpr();
+                    else
+                        error("`typeof` operator expects an expression or type name");
```

Comment:
The grammar is:
```
type-specifier:
    typeof-specifier

typeof-specifier:
    typeof ( expression )
    typeof ( type-name )
```

This check is too narrow, it's best to defer it to semantic.